rpc/core/tx.go

```diff
@@ -0,0 +1,81 @@
+package core
+
+import (
+	"fmt"
+
+	abci "github.com/tendermint/abci/types"
+	ctypes "github.com/tendermint/tendermint/rpc/core/types"
+	"github.com/tendermint/tendermint/state/tx/indexer"
+	"github.com/tendermint/tendermint/types"
+)
+
+func Tx(hash []byte, height, index int, prove bool) (*ctypes.ResultTx, error) {
+
+	// if index is disabled, we need a height
+	_, indexerDisabled := txIndexer.(*indexer.Null)
+	if indexerDisabled && height == 0 {
+		return nil, fmt.Errorf("TxIndexer is disabled. Please specify a height to search for the tx by hash or index")
+	}
+
+	// hash and index must not be passed together
+	if len(hash) > 0 && index != 0 {
+		return nil, fmt.Errorf("Invalid args. Only one of hash and index may be provided")
+	}
+
+	// results
+	var txResult abci.ResponseDeliverTx
+	var tx types.Tx
+
+	// if indexer is enabled and we have a hash,
+	// fetch the tx result and set the height and index
+	if !indexerDisabled && len(hash) > 0 {
+		r, err := txIndexer.Tx(hash)
+		if err != nil {
+			return nil, err
+		}
+
+		if r == nil {
+			return &ctypes.ResultTx{}, fmt.Errorf("Tx (%X) not found", hash)
+		}
+
+		height = int(r.Height) // XXX
+		index = int(r.Index)
+		txResult = r.DeliverTx
+	}
+
+	// height must be valid
+	if height <= 0 || height > blockStore.Height() {
+		return nil, fmt.Errorf("Invalid height (%d) for blockStore at height %d", height, blockStore.Height())
+	}
+
+	block := blockStore.LoadBlock(height)
```

Comment:
> Do we need to load a block if we have the txid?

Yes because txs aren't stored with the result since then they'd be stored twice. So if you want the actual tx, you need to load a block.

> I feel that we're overloading this function... I'd rather we only allowed calling this function if tx indexing is enabled.

I agree - as I was saying on call today. The idea here was that if we don't have the indexer but we do have the txid, we can still search the block for it. That still seems like useful functionality. Maybe we should separate into two calls tho: `/tx` and `/tx_result`. The former will always work - if the indexer is active, it can fetch the height/index and get the tx from the block, else it requires a height be sent as an arg and will search the block for it. The latter will only work if the indexer is active and will only return the result, no tx. Or we could add a boolean `tx` arg to `tx_result` to specify if you want the `tx` bytes too